--- a/scripts/trylink
+++ b/scripts/trylink
@@ -93,18 +93,14 @@
     SORT_SECTION=""
 fi

-START_GROUP="-Wl,--start-group"
-END_GROUP="-Wl,--end-group"
+START_GROUP=""
+END_GROUP=""
 INFO_OPTS() {
-	echo "-Wl,--warn-common -Wl,-Map,$EXE.map -Wl,--verbose"
+	: # wasm-ld: flags not supported
 }

-# gold may not support --sort-common (yet)
-SORT_COMMON="-Wl,--sort-common"
-if ! check_cc "-Wl,--sort-common"; then
-    echo "Your linker does not support --sort-common"
-    SORT_COMMON=""
-fi
+# wasm-ld does not support --sort-common
+SORT_COMMON=""

 # Static linking against glibc produces buggy executables
 # (glibc does not cope well with ld --gc-sections).
