diff --git a/Src/params.c b/Src/params.c
index 213b6f4..62914b0 100644
--- a/Src/params.c
+++ b/Src/params.c
@@ -3659,12 +3659,17 @@ unsetparam_pm(Param pm, int altflag, int exp)
 		altpm->level = 1;
 	    }
 	    unsetparam_pm(altpm, 1, exp);
+	    /* This resets pm to empty */
+	    pm->node.flags |= PM_UNSET;
+	    /* so we must repeat this */
 	}
 
 	zsfree(altremove);
-	if (!(pm->node.flags & PM_SPECIAL))
+	if (!(pm->node.flags & PM_SPECIAL)) {
 	    pm->gsu.s = &stdscalar_gsu;
+	    pm->node.flags &= ~PM_ARRAY;
+	}
     }
 
     /*
